File: cpp/includes/graph.cpp
#include <string>
#include "includes/team"
#include "includes/adj_list"
#include "includes/borie_rating_algorithm"
using namespace std;

class Graph
{
    private:
        /*
            TODO: These aren't right...
            Possibly replace Dict and string[] with std::map<>
        */
        Dict teams;
        string[] array;
        int size;
        RatingAlgorithm algorithm;

    public:
        Graph(void)
        {
            this->teams = {};
            this->array = [];
            this->size = 0;
            this->algorithm = RatingAlgorithm();
        }

        void buildGraph(reader)
        {
            /*
            file = reader.readFile()

            for line in file:
                game = reader.parseLine(line)

            if game:
                scoreDiff1 = eval(game.group(2)) - eval(game.group(4))
                scoreDiff2 = 0 - scoreDiff1
                scoreDiff1 = this.algorithm.improveScores(scoreDiff1)
                scoreDiff2 = this.algorithm.improveScores(scoreDiff2)
                this.fillGraph(game.group(1),game.group(3),scoreDiff1,scoreDiff2)
            */
            print("Done")
        }


        void fillGraph(string team1, string team2, int scoreDiff1, int scoreDiff2)
        {
            if (team1 not in this->teams.keys()) { this->addTeam(team1); }
            this->addGame(team1,team2,scoreDiff1);
            if (team2 not in this->teams.keys()) { this->addTeam(team2); }
            this->addGame(team2,team1,scoreDiff2);
        }

        void addTeam(Team team)
        {
            A = AdjList();
            A.insertHead(team);
            this->array.append(A);
            this->teams[team] = this->size;
            this->size += 1;
        }

        void addGame(string team1, string team2,int scoreDiff)
        {
            arrayIndex = this.getTeam(team1);
            this->array[arrayIndex].insert(team2,float(scoreDiff));
        }

        int getSize(this) { return this->size; }
        Team getTeam(this,name) { return this->teams[name]; }

        void rateTeams(this)
        {
            println('Calculating Ratings...');
            this->array = this.algorithm.rateTeams(this->array, this->teams);
            println('Done');
        }

        Dict getTeams(this) { return this->teams; }
        string[] getTeamNames(this) { return this->array; }
}

File: python/includes/graph.py
from includes.team import Team
from includes.adj_list import AdjList
from includes.borie_rating_algorithm import RatingAlgorithm

class Graph():

    def __init__(self):
        self._teams = {}
        self._array = []
        self._size = 0
        self.algorithm = RatingAlgorithm()

    def buildGraph(self, reader):
        file = reader.readFile()

        games = reader.parseFile(file)

        if games:
            for game in games:
                scoreDiff1 = eval(game[1]) - eval(game[3])
                scoreDiff2 = 0 - scoreDiff1
                scoreDiff1 = self.algorithm.improveScores(scoreDiff1)
                scoreDiff2 = self.algorithm.improveScores(scoreDiff2)
                self.fillGraph(game[0],game[2],scoreDiff1,scoreDiff2)
        print("Done")


    def fillGraph(self,team1,team2,scoreDiff1,scoreDiff2):
        if team1 not in self._teams.keys(): self.addTeam(team1)
        self.addGame(team1,team2,scoreDiff1)
        if team2 not in self._teams.keys(): self.addTeam(team2)
        self.addGame(team2,team1,scoreDiff2)

    def addTeam(self,team):
        A = AdjList()
        A.insertHead(team)
        self._array.append(A)
        self._teams[team] = self._size
        self._size += 1

    def addGame(self,team1,team2,scoreDiff):
        arrayIndex = self.getTeam(team1)
        self._array[arrayIndex].insert(team2,float(scoreDiff))

    def getSize(self): return self._size
    def getTeam(self,name): return self._teams[name]

    def rateTeams(self):
        print('Calculating Ratings...')
        self._array = self.algorithm.rateTeams(self._array, self._teams)
        print('Done')

    def getTeams(self): return self._teams
    def getTeamNames(self): return self._array